cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp

```diff
@@ -83,78 +84,112 @@ static NumericTablePtr get_initial_centroids(const dal::backend::context_gpu& ct
+
+        Float prev_objective_function = de::limits<Float>::max();
+        std::int64_t iter;
+        sycl::event centroids_event;
+
+        auto updater = cluster_updater<Float>{}
+                           .set_cluster_count(cluster_count)
+                           .set_part_count(part_count)
+                           .set_queue(queue)
+                           .set_data(arr_data);
+        updater.allocate_buffers();
+
+        for (iter = 0; iter < max_iteration_count; iter++) {
+            updater.set_initial_centroids(iter == 0 ? arr_initial : arr_centroids);
+            auto [objective_function, update_clusters_event] =
+                updater.update(arr_centroids,
+                               arr_distance_block,
+                               arr_closest_distances,
+                               arr_objective_function,
+                               arr_labels,
+                               { centroids_event });
+            centroids_event = update_clusters_event;
+            if (accuracy_threshold > 0 &&
+                objective_function + accuracy_threshold > prev_objective_function) {
+                iter++;
+                break;
+            }
+            prev_objective_function = objective_function;
+        }
+        auto assign_event =
+            kernels_fp<Float>::template assign_clusters<pr::squared_l2_metric<Float>>(
+                queue,
+                arr_data,
+                arr_centroids,
+                block_size_in_rows,
+                arr_labels,
+                arr_distance_block,
+                arr_closest_distances,
+                { centroids_event });
+        kernels_fp<Float>::compute_objective_function(queue,
+                                                      arr_closest_distances,
+                                                      arr_objective_function,
+                                                      { assign_event });
         return train_result<task::clustering>()
-            .set_labels(
-                dal::detail::homogen_table_builder{}.reset(arr_labels, row_count, 1).build())
-            .set_iteration_count(static_cast<std::int64_t>(arr_iter_count_host[0]))
-            .set_objective_function_value(static_cast<double>(arr_obj_func_host[0]))
+            .set_labels(dal::homogen_table::wrap(arr_labels.flatten(queue), row_count, 1))
+            .set_iteration_count(iter)
+            .set_objective_function_value(arr_objective_function.to_host(queue).get_data()[0])
             .set_model(model<task::clustering>().set_centroids(
-                dal::detail::homogen_table_builder{}
-                    .reset(arr_centroids, cluster_count, column_count)
-                    .build()));
+                dal::homogen_table::wrap(arr_centroids.flatten(queue),
+                                         cluster_count,
+                                         column_count)));
```

Comment:
Declare separate variable for model, this is barely readable

Reply:
Done